packages/ses/src/module-proxy.js

```diff
@@ -0,0 +1,137 @@
+// Compartments need a mechanism to link a module from one compartment
+// to another.
+// The procedure is to use `compartment.module(specifier)` to obtain the module
+// exports namespace from one compartment, possibly before importing or even
+// merely loading module, and threading it into the module map of another
+// compartment.
+// For this to be possible, it is necessary to model the module exports
+// namespace as a proxy that will treat all access to those exported properties
+// as reference errors until the properties of the actual module are known.
+// This provides the mechanism for modeling the public exports proxy
+// and eventually connecting it to to the proxied exports.
+
+import { makeAlias } from './module-load.js';
+
+const { create, freeze } = Object;
+// q, as in quote, for error messages.
+const q = JSON.stringify;
+
+// `deferExports` creates a module's exports proxy, proxied exports, and
+// activator.
+// A `Compartment` can create a module for any module specifier, regardless of
+// whether it is loadable or executable, and use that object as a token that
+// can be fed into another compartment's module map.
+// Only after the specified module has been analyzed is it possible for the
+// module namespace proxy to behave properly, so it throws exceptions until
+// after the compartment has executed the module.
```

Comment:
Did we give up any useful functionality by waiting until execution, rather than after analysis and before execution?

To be precise (as you are elsewhere), rather than say "until after ... has executed ..." say "until ... had begun executing ..."

Reply:
This has changed and I believe the current prose is accurate. The module proxy remains inactive until the module has run to completion. We could reveal the exports when the module begins executing, but that would a more invasive design change we previously discussed: creating an intermediate proxy target that is just a bank of accessors for all of the module’s expected exports including those inherited through `import * from` directives, then ensuring that all their TDZs are faithfully modeled if someone uses the module namespace object between the beginning and end of a module’s execution. 

Before we implement top-level-await, I don’t believe that period is observable to anyone holding an exports namespace proxy. There are two reasons and means for obtaining the namespace proxy. One is to link compartments and generally will occur from outside the compartment. This class of users has no expectations about the specific shape of the exports.

The entire class of users who obtain a module exports namespace will generally obtain it from `import` or `importNow`, either inside or outside the compartment. Users of `import` will not be able to observe the namespace until after it finishes executing, gets sealed, activated, and resolves the promise returned by `import`. Users of `importNow` will not be able to expect to observe the namespace if `importNow` does not throw an error, which can only occur the turn after the promise returned by `import` resolves.

I conclude that there’s no expectation for the namespace to be usable between when a module begins executing and ends executing. We could choose to make the namespace observable between the begin and end of execution after we introduce top-level-await, but that itself would introduce a hazard, since the user would only be able to predict whether the namespace is useable by synchronizing with `import` or `importNow`.

Reply:
I suppose it is possible to obtain a module namespace object through `compartment.module`, then pass that into the compartment with `evaluate` or by passing it in through an endowment. In that case, it might be possible to orchestrate a scenario where the module would synchronously emit callbacks to mark progress of execution of the module, such that the holder of the namespace object could rely upon various stages of that progress.

I think it might be better to preclude that kind of cleverness.

Reply:
I’ve noticed that my very precise error messages are not accurate. I’ll change these, for example:

```js
          throw new TypeError(
            `Cannot check property ${q(
              name,
            )}, the module has not yet finished executing`,
         );
```

Reply:
activate is called at the end of the `import` function, which is itself called by the module functor when the module begins execution, but before executing any use code. I think "had begun executing" is faithful to your implementation.